crates/accelerate/src/unitary_synthesis.rs

```diff
@@ -0,0 +1,1093 @@
+                            params.iter().zip(target_op.params.iter()).all(|(p1, p2)| {
+                                p1.is_close(py, p2, 1e-10)
+                                    .expect("Unexpected parameter expression error.")
+                            })
+                    } else {
+                        false
+                    };
+                    let is_parametrized = target_op
+                        .params
+                        .iter()
+                        .any(|param| matches!(param, Param::ParameterExpression(_)));
+                    if target_op.operation.name() == inst_name
+                        && (is_parametrized || are_params_close)
+                    {
+                        match target[name].get(Some(inst_qubits)) {
+                            Some(Some(props)) => gate_fidelities.push(
+                                1.0 - props.error.unwrap_or(0.0)
+                            ),
+                            _ => gate_fidelities.push(1.0),
+                        }
+                        break;
+                    }
+                }
+                Ok(())
+            }
+            Err(_) => {
+                Err(QiskitError::new_err(
+                    format!("Encountered a bad synthesis. Target has no instruction {inst_name:?} on qubits {inst_qubits:?}.")
+                ))
+            }
+        }
+    };
+
+    for (inst_name, inst_params, inst_qubits) in synth_circuit {
+        let _ = score_instruction(&inst_name, &inst_params, &inst_qubits);
+    }
+    1.0 - gate_fidelities.into_iter().product::<f64>()
+}
+
+// This is the outer-most run function. It is meant to be called from Python
+// in `UnitarySynthesis.run()`.
+#[pyfunction]
+#[pyo3(name = "run_default_main_loop")]
+fn py_run_main_loop(
+    py: Python,
+    dag: &mut DAGCircuit,
+    qubit_indices: &Bound<'_, PyList>,
+    min_qubits: usize,
+    target: &Target,
+    coupling_edges: &Bound<'_, PyList>,
+    approximation_degree: Option<f64>,
+    natural_direction: Option<bool>,
+) -> PyResult<DAGCircuit> {
+    let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py);
+    let dag_to_circuit = imports::DAG_TO_CIRCUIT.get_bound(py);
+
+    // Run synthesis recursively over control flow ops, mapping qubit indices
+    let node_ids: Vec<NodeIndex> = dag.op_nodes(false).collect();
+    for node in node_ids {
+        let NodeType::Operation(inst) = &dag.dag()[node] else {
+            panic!("DAG node must be an instruction")
+        };
+        if !inst.op.control_flow() {
+            continue;
+        }
+        let OperationRef::Instruction(py_inst) = inst.op.view() else {
+            panic!("Control flow op must be an instruction")
+        };
+        let raw_blocks: Vec<PyResult<Bound<PyAny>>> = py_inst
+            .instruction
+            .getattr(py, "blocks")?
+            .bind(py)
+            .iter()?
+            .collect();
+        let mut new_blocks = Vec::with_capacity(raw_blocks.len());
```

Comment:
As of #13056 you can use the method call `PackedOperation::blocks` to retrieve the blocks from python-space. Now, the return type for this will be `CircuitData` which has a more limited API than `QuantumCircuit`. Since you seem to be using `replace_blocks` further down, you'd need to convert these back to QuantumCircuit which happens to be a lossy conversion, so maybe you can keep these lines in, and change it in a follow-up.

Reply:
Yes, I didn't use `PackedOperation::blocks` because I would need the `QuantumCircuit` down the line. This is also why I wasn't convinced by using the rust-space `dag_to_circuit` method, because I would still need to convert back and forth between the rust and the python objects a couple of times.